### PR TITLE
Add CI for Unreal Engine 5.4

### LINKED ISSUE #1400: Add CI support for Unreal Engine 5.4
For our github action builds, we need new jobs for Unreal Engine 5.4.

- Support all the existing platforms `Windows, Android, Linux, macOS, iOS`.
- Generate new artifacts, ex. `CesiumForUnreal-54-<platform>-<branch>`
- Potentially remove support for 5.1, as we traditionally only support 3 versions at a time.

The cesium-unreal plugin should build successfully as of [this PR](https://github.com/CesiumGS/cesium-unreal/pull/1396)

### PR DESCRIPTION
**Pull request by kring**
This builds on #1385 because main is broken.

Closes #1400

### CODE REVIEW

File: .github/actions/install-unreal-linux/action.yml [hunk 1]
```diff
@@ -27,5 +27,6 @@ runs:
       shell: bash
       run: |
         mkdir -p /mnt/work/${{ inputs.unreal-program-name }}
-        unzip -q ./ue.zip -d /mnt/work/${{ inputs.unreal-program-name }}
+        # Selectively extract files in order to save disk space
```
Comment (csciguy8) [inline .github/actions/install-unreal-linux/action.yml:30]: Looks like a straight forward optimization..

... but if there are unneeded files in the linux ue zip, could just make our own .zip on s3 (like the others) that already have the unnecessary files pruned out. Would save on download bandwidth and the extra unzip complexity.

**Review by csciguy8 — COMMENTED** [review]
Thanks @kring!

Good changes to get CI up and running. It looks like things are building and artifacts are being generated, but I did notice a couple of issues that may / may not be solvable...

1) Build times for the new 5.4 version seem to be slower. The Windows 5.4 build is ~2x slower than the 5.3 build. Android54, linux54, iOS54 seem to be slight slower too. macOS54 is about the same

2) Artifact size for 5.4 seems to be ~2x larger than 5.3. This is a bit concerning.

**Comment by csciguy8** [comment]
> 2\. Artifact size for 5.4 seems to be ~2x larger than 5.3. This is a bit concerning.

Good news! I diff'd the artifacts and discovered that the problem seems to be in the `Intermediate` folder. Usually we see `Module.CesiumRuntime.X.cpp.o` files in there. But with the new 5.4 build, we're seeing additional .o files (non-unity build).

image

Bad news, not sure what we did to cause this. From the looks of it, we build 5.4 just like we build 5.3. Worst case we just clean up these extra files before creating the artifact?

**Comment by csciguy8** [comment]
Also noticed that 5.4 packaging failed...

```
[11/13] Compile [x64] Module.TestsProject.cpp
[12/13] Link [x64] TestsProject.exe
CesiumGltfReader.lib(GltfReader.obj) : error LNK2005: stbir_resize already defined in Module.ImageCore.cpp.obj
CesiumGltfReader.lib(GltfReader.obj) : error LNK2005: stbir_resize_uint8_srgb already defined in Module.ImageCore.cpp.obj
   Creating library D:\a\cesium-unreal\cesium-unreal\TestsProject\Binaries\Win64\TestsProject.lib and object D:\a\cesium-unreal\cesium-unreal\TestsProject\Binaries\Win64\TestsProject.exp
D:\a\cesium-unreal\cesium-unreal\TestsProject\Binaries\Win64\TestsProject.exe : fatal error LNK1169: one or more multiply defined symbols found
```

https://github.com/CesiumGS/cesium-unreal/actions/runs/8893486883/job/24432801834

File: .github/actions/install-unreal-linux/action.yml (hunk 1 above)
Comment (kring) [reply .github/actions/install-unreal-linux/action.yml:30]: This isn't so much an optimization as a desperate attempt to get it to work at all. The GH Actions Linux runner doesn't have enough disk space to extract the whole thing. Yes, we could optimize the ZIP from here.

**Comment by csciguy8** [comment]
> [11/13] Compile [x64] Module.TestsProject.cpp
> [12/13] Link [x64] TestsProject.exe
> CesiumGltfReader.lib(GltfReader.obj) : error LNK2005: stbir_resize already defined in Module.ImageCore.cpp.obj
> CesiumGltfReader.lib(GltfReader.obj) : error LNK2005: stbir_resize_uint8_srgb already defined in Module.ImageCore.cpp.obj
>    Creating library D:\a\cesium-unreal\cesium-unreal\TestsProject\Binaries\Win64\TestsProject.lib and object D:\a\cesium-unreal\cesium-unreal\TestsProject\Binaries\Win64\TestsProject.exp
> D:\a\cesium-unreal\cesium-unreal\TestsProject\Binaries\Win64\TestsProject.exe : fatal error LNK1169: one or more multiply defined symbols found

`Module.ImageCore.cpp.obj` is an obj file in the 5.4 engine source... which hints that stb symbols were newly defined here in 5.4, and we need to work around this somehow
`UE_5.4\Engine\Intermediate\Build\Win64\x64\UnrealGame\Development\ImageCore`

.

**Comment by kring** [comment]
> Module.ImageCore.cpp.obj is an obj file in the 5.4 engine source... which hints that stb symbols were newly defined here in 5.4, and we need to work around this somehow

Yep. We'll probably have to modify the cesium-native build to optionally not include those `sbir` functions. We may also have to add ImageCore or similar to CesiumRuntime's .build.cs so that non-packaged builds don't break.

**Comment by csciguy8** [comment]
I think I have a fix for the packaging test failure (stb linker error). It involved a change to cesium-native so I committed that.

In theory, both cesium-native and cesium-unreal should now be able to build (crossing fingers).

**Comment by kring** [comment]
If your fix works (it looks promising), then perhaps we can ship UE 5.4 support today after all. I see you cancelled the build, though? Was that because you didn't think it would work, or you were just trying not to use build time during the release?

**Comment by csciguy8** [comment]
> Was that because you didn't think it would work, or you were just trying not to use build time during the release?

It was because I didn't think it would work. I saw an error, then fixed it, then reupdated cesium-native in this PR.

Should be good to go now. Just waiting for CI results for this PR.

We now depend on [cesium-native PR #871](https://github.com/CesiumGS/cesium-native/pull/871)

**Pull request merged**